### ISSUE TITLE
Table View Switch Accessories Not Keeping State

### ISSUE THREAD

**Issue by evansobkowicz** [clearsightstudio/ProMotion#625]
When scrolling through a table view where every cell has a switch accessory, and the list is longer than the device's screen, the switches lose their 'on' state when scrolled off the screen and then back on. 

Is there a way to fix this either in my app or in ProMotion? Any advice?

Thanks!

**Comment by markrickert** [comment]
What's likely happening here is that the cells are being reused.

Try specifying a different `cell_identifier` for each table cell that has an accessory and it shouldn't do this any more. It could also have to do with the state management when it recreates the cell. Make sure you're storing the state of the accessory in an instance var so that the cell knows what to "default" it to when the cell is recreated.

**Comment by jamonholmgren** [comment]
I'm not sure the `cell_identifier` will help in this case, if I'm understanding the problem. Basically, Evan is saying that Cell 1 is switched "off", he switches it "on", scrolls down, back up, and Cell 1 is "off" again.

**Comment by evansobkowicz** [comment]
Yes, @jamonholmgren is right about the problem. I don't think cell_identifier fixes the problem. Would a demo app be helpful?

**Comment by jamonholmgren** [comment]
When I'm thinking this through, I think we need a way to reload the data without reloading the table entirely. something like:

```ruby
# in cell hash:
{
  accessory: {
    view: :switch,
    action: :switch_accessory,
    arguments: { item: item },
  }
},

def accessory_switched(args={})
  args[:item].switched_on = !args[:item].switched_on
  update_table_data
end
```

This will have an undesirable full table reload, so it might be a good idea to add another option to `update_table_data` which prevents that:

```ruby
update_table_data(reload_table: false)
```

**Comment by jamonholmgren** [comment]
Demo would help, @evansobkowicz, so we could see it in action.

**Comment by evansobkowicz** [comment]
Here's a demo app showing the issue. 'true' values set initially don't have this problem, but values switched on after load don't keep their state on scroll. When they are switched on again, their value is removed from the 'selected_states' array. Let me know if you need any more explanation.

https://github.com/evansobkowicz/TableViewSwitchBug

**Comment by evansobkowicz** [comment]
@jamonholmgren Just wondering if there's anything else you need from me or that I can help with!

**Comment by forrestgrant** [comment]
I think this is because every time a Cell appears on screen (regardless of a previous appearance), UITableViewDelegate is calling: `def tableView(table_view, willDisplayCell: table_cell, forRowAtIndexPath: index_path)` which reads from tabledata, even though the switch changed state, it hasn't persisted to tabledata.

Working on a sample.

**Comment by forrestgrant** [comment]
This is pretty ugly, but works:

```ruby
  def accessory_switched(args = {})
    @selected_states.include?(args[:name]) ? remove_from_selected(args) : add_to_selected(args)
    self.promotion_table_data.data[0][:cells].each do |cell|
      if cell[:title] == args[:name]
        cell[:accessory][:value] = @selected_states.include?(args[:name])
      end
    end
  end
```

**Comment by jamonholmgren** [comment]
@forrestgrant is right. What we need to do is provide an `update_table_data reload_table: false` method so you can reload table_data without reloading the table. It shouldn't be hard; just don't run these two lines if it's set to false: https://github.com/clearsightstudio/ProMotion/blob/version-2.3/lib/ProMotion/table/table.rb#L96-L99

**Comment by jamonholmgren** [comment]
I've run into this issue, and it doesn't have anything to do with what I thought it did. Working on it now.

cc @kevinvangelder

**Comment by augustosamame** [comment]
Hi. This issue just bit me. As soon as I scroll, my switches lose their ON status. I don't understand if this issue should have been fixed (as it is closed), or if I need to use the workaround indicated in the comments.

In that case, is this 

`update_table_data(reload_table: false)`

working as well?

thanks in advance for any help

**Comment by jamonholmgren** [comment]
I'm not sure @augustosamame -- it's been a year and a half since I ran into this. Definitely update if you figure out a solution.

**Comment by augustosamame** [comment]
Well, I tried it and 
1) the workaround works and switches are now keeping state
2) I have tried both update_table_data and update_table_data(reload_table: false) and I can't see any difference, although my table is small so maybe that's the reason. At least it did not throw an error.

Taking a look at the source code:

https://github.com/infinitered/ProMotion/blob/0be19ab6aecd401e22315e38b8f3b927367e4f64/lib/ProMotion/table/table.rb

it seems the `(reload_table: false)`  option has not been implemented.

it doesn't seem an issue with small tables though so I think the workaround is fine.
